model.py

```diff
@@ -6616,6 +6620,98 @@ def update_availability(
+        should suffice until more accurate information can be
+        obtained.
+
+        No CirculationEvent is created until `update_availability` is
+        called.
+
+        Events must be processed in chronological order. Any event
+        that happened than `LicensePool.last_checked` is ignored, and
+        calling this method will update `LicensePool.last_checked` to
+        the time of the event.
+
+        :param event_type: A CirculationEvent constant representing the
+        type of change that was seen.
+
+        :param event_date: A datetime corresponding to when the 
+        change was seen.
+
+        :param delta: The magnitude of the change that was seen.
+
+        """
+        if event_date != CirculationEvent.NO_DATE and self.last_checked and event_date < self.last_checked:
+            # This is an old event and its effect on availability has
+            # already been taken into account.
+            return
+
+        if self.last_checked and event_date == CirculationEvent.NO_DATE:
+            # We have a history for this LicensePool and we don't know
+            # where this event fits into that history. Ignore the
+            # event.
+            return
+
+        (new_licenses_owned, new_licenses_available, 
+         new_licenses_reserved, 
+         new_patrons_in_hold_queue) = self._calculate_change_from_one_event(
+             event_type, delta
+         )
+        self.update_availability(
+            new_licenses_owned, new_licenses_available, 
+            new_licenses_reserved, new_patrons_in_hold_queue,
+            analytics=analytics, as_of=event_date
+        )
+
+    def _calculate_change_from_one_event(self, type, delta):
+        new_licenses_owned = self.licenses_owned
+        new_licenses_available = self.licenses_available
+        new_licenses_reserved = self.licenses_reserved
+        new_patrons_in_hold_queue = self.patrons_in_hold_queue
+
+        def deduct(value):
+            # It's impossible for any of these numbers to be
+            # negative.
+            return max(value-delta, 0)
+
+        CE = CirculationEvent
+        added = False
+        if type == CE.DISTRIBUTOR_HOLD_PLACE:
+            new_patrons_in_hold_queue += delta
+        elif type == CE.DISTRIBUTOR_HOLD_RELEASE:
+            new_patrons_in_hold_queue = deduct(new_patrons_in_hold_queue)
+        elif type == CE.DISTRIBUTOR_CHECKIN:
+            new_licenses_available += delta
+        elif type == CE.DISTRIBUTOR_CHECKOUT:
+            new_licenses_available = deduct(new_licenses_available)
+        elif type == CE.DISTRIBUTOR_LICENSE_ADD:
+            new_licenses_owned += delta
+        elif type == CE.DISTRIBUTOR_LICENSE_REMOVE:
+            new_licenses_owned = deduct(new_licenses_owned)
+        elif type == CE.DISTRIBUTOR_AVAILABILITY_NOTIFY:
+            new_patrons_in_hold_queue = deduct(new_patrons_in_hold_queue)
+            new_licenses_reserved += delta
```

Comment:
Would licenses_available also change in this case? Or would the previous checkin event have had delta 0?